pallets/connectors-gateway/src/lib.rs

```diff
@@ -0,0 +1,273 @@
+// Copyright 2023 Centrifuge Foundation (centrifuge.io).
+//
+// This file is part of the Centrifuge chain project.
+// Centrifuge is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version (see http://www.gnu.org/licenses).
+// Centrifuge is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use core::fmt::Debug;
+
+use cfg_traits::connectors::{Codec, InboundQueue, OutboundQueue, Router as DomainRouter};
+use cfg_types::domain_address::{Domain, DomainAddress};
+use codec::{EncodeLike, FullCodec};
+use frame_support::{dispatch::DispatchResult, pallet_prelude::*};
+use frame_system::pallet_prelude::OriginFor;
+pub use pallet::*;
+use sp_std::convert::TryInto;
+
+use crate::weights::WeightInfo;
+
+mod origin;
+pub use origin::*;
+
+pub mod weights;
+
+#[cfg(test)]
+mod mock;
+
+#[cfg(test)]
+mod tests;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub (super) trait Store)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::origin]
+	pub type Origin = GatewayOrigin;
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// The origin type.
+		type RuntimeOrigin: Into<Result<GatewayOrigin, <Self as frame_system::Config>::RuntimeOrigin>>
+			+ From<GatewayOrigin>;
+
+		/// The event type.
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		/// The LocalOrigin ensures that some calls can only be performed from a
+		/// local context i.e. a different pallet.
+		type LocalOrigin: EnsureOrigin<
```

Comment:
Maybe nit: To me, the prefix `local` is a bit misleading here as we have been referring to `Centrifuge` as the local domain. However, in the current implementation we error out if its not an EVM domain.

Reply:
Yea, by `local` we mean a different pallet. Maybe call it `InnerOrigin` or something better? ^^

Reply:
Too many points of view for the things in connectors. We need a connector's relativist theory 😆